### PR TITLE
Implement support for sample rate divider in multireader

### PR DESCRIPTION
**Pull request by metod-c**
## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Checklist:

- [ ] Pull request title reflects its content
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# Description:

### CODE REVIEW

File: core/opendaq/reader/tests/test_multi_reader.cpp [hunk 1]
```diff
@@ -1813,4 +1813,68 @@ TEST_F(MultiReaderTest, StartOnFullUnitOfDomain)
 
     ASSERT_THAT(time[1], ElementsAreArray(time[0]));
     ASSERT_THAT(time[2], ElementsAreArray(time[0]));
-}
\ No newline at end of file
+}
+
+TEST_F(MultiReaderTest, SampleRateDivider)
```
Comment (JakaMohorkoDS) [inline core/opendaq/reader/tests/test_multi_reader.cpp:1818]: A test of an expected error when the signals do not have an integer divider would be a good addition.

**Review by JakaMohorkoDS — COMMENTED** [review]
Lgtm, missing a way of getting dividers at the interface level, comments for the `read` method, that it allows for dividers, and a fail-state test case.

File: core/opendaq/reader/include/opendaq/multi_reader_impl.h [hunk 2]
```diff
@@ -83,6 +83,7 @@ class MultiReaderImpl : public ImplementationOfWeak<IMultiReader, IReaderConfig,
     template <typename T>
     static bool ListElementsHaveSameType(const ListPtr<IBaseObject>& list);
     static bool CheckPreconditions(const ListPtr<IBaseObject>& list);
+    void updateCommonSampleRateAndDividers();
```
[the remark this answers is not in the document]
Comment (dejancrn) [reply core/opendaq/reader/include/opendaq/multi_reader_impl.h:86]: We even don't know the sampling rate of the signals from multi reader interface. We have to extract the sampling rate of the signal from it's descriptor attribute, which may in some cases be different than the descriptor that arrives through event packet. So sample rate dividers are no different than sample rates. What I would like to see is a dictionary of <ISignal, IDataDescriptor> on the multi reader interface. 

I agree it is nice to have a function that returns dividers, but that should be helper function that calculates the information from <ISignal, IDataDescriptor> interface.

But this is probably not a part of this PR.

**Review by dejancrn — APPROVED** [review]

File: core/opendaq/reader/include/opendaq/multi_reader_impl.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (JakaMohorkoDS) [reply core/opendaq/reader/include/opendaq/multi_reader_impl.h:86]: I don't fully agree here. The `read` call, at the moment, gives `count` as an [out] param. This count is the maximum samples read if I'm not mistaken. From the `read` call we can never really know how many samples we got for each individual signal - the dividers would make inferring that a lot easier.

The documentation and examples at the moment don't mention how an application is supposed to read data from multiple signals - right now users would assume that the `count` value holds true for all signals. If nothing else, we need somewhere a note/comment/how-to that warns the users that the read sample count is very likely going to be different per-signal.

**Pull request merged**